File: [CCS LABS] Windows Forms Dynamically Writing to the Background Desktop - Part 2/[C#][VB]-[CCS LABS] Windows Forms Dynamically Writing to the Background Desktop - Part 2/C# and VB/C#,VB/BGInfo2/DelegateExpansion.cs
using System;
using System.Windows.Forms;

namespace BGInfo2
{
    public static class DelegateExpansion
    {
        // Prevent CrossThreadException by invoking delegate through target control's thread.
        public static object CrossInvoke(this Delegate delgt, object sender, EventArgs e)
        {
            if (delgt.Target is Control && ((Control)delgt.Target).InvokeRequired)
            {
                return ((Control)delgt.Target).Invoke(delgt, new object[]
                {
                    sender,
                    e
                });
            }
            return delgt.Method.Invoke(delgt.Target, new object[]
            {
                sender,
                e
            });
        }
    }
}

File: [CCS LABS] Windows Forms Dynamically Writing to the Background Desktop -BGInfo2/[VB]-[CCS LABS] Windows Forms Dynamically Writing to the Background Desktop -BGInfo2/VB/BGInfo2.ConvertedToVBNet/DelegateExpansion.vb
Imports System.Windows.Forms

Public NotInheritable Class DelegateExpansion
	Private Sub New()
	End Sub
	' Prevent CrossThreadException by invoking delegate through target control's thread.
	
	Public Shared Function CrossInvoke(delgt As [Delegate], sender As Object, e As EventArgs) As Object
		If TypeOf delgt.Target Is Control AndAlso DirectCast(delgt.Target, Control).InvokeRequired Then
			Return DirectCast(delgt.Target, Control).Invoke(delgt, New Object() {sender, e})
		End If
		Return delgt.Method.Invoke(delgt.Target, New Object() {sender, e})
	End Function
End Class
